burn-tensor/src/tensor/api/base.rs

```diff
@@ -168,6 +173,93 @@ where
+    /// appropriate formatting.
+    ///
+    /// # Arguments
+    ///
+    /// * `acc` - A mutable reference to a `String` used as an accumulator for the formatted output.
+    /// * `depth` - The current depth of the tensor dimensions being processed.
+    /// * `multi_index` - A mutable slice of `usize` representing the current indices in each dimension.
+    /// 
+    /// # Example
+    /// 
+    /// ```rust
+    /// let data = Data::from([[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]);
+    /// let tensor: burn_tensor::Tensor<TestBackend, 3, burn_tensor::Int> = Tensor::from_data(data);
+    /// let mut acc = String::new();
+    /// let mut multi_index = vec![0; 3];
+    /// tensor.display_recursive(&mut acc, 0, &mut multi_index);
+    /// assert_e!(acc, "[[[1, 2, 3], [4, 5, 6]], [[7, 8, 9], [10, 11, 12]]]");
+    /// ```
+    fn display_recursive(&self, acc: &mut String, depth: usize, multi_index: &mut [usize]) {
+        if depth == 0 {
+            acc.push('[');
+        }
+    
+        if depth == self.dims().len() - 1 {
+            // if we are at the innermost dimension, just push its elements into the accumulator
+            for i in 0..self.dims()[depth] {
+                if i > 0 {
+                    acc.push_str(", ");
+                }
+                multi_index[depth] = i;
+                let range: [std::ops::Range<usize>; D] = core::array::from_fn(|i| {
+                    multi_index[i].clone()..multi_index[i].clone() + 1
+                });
+                let elem = self.clone().index(range).to_data().value[0];
+                acc.push_str(&format!("{:?}", elem));
+            }
+        } else {
+            // otherwise, iterate through the current dimension and recursively display the inner tensors
+            for i in 0..self.dims()[depth] {
+                if i > 0 {
+                    acc.push_str(", ");
+                }
+                acc.push('[');
+                multi_index[depth] = i;
+                self.display_recursive(acc, depth + 1, multi_index);
+                acc.push(']');
+            }
+        }
+    
+        if depth == 0 {
+            acc.push(']');
+        }
+    }
+}
+
+/// Pretty print Int tensors
+impl<B, const D: usize> std::fmt::Display for Tensor<B, D, Int>
+where
+    B: Backend,
+    B::IntElem: std::fmt::Display,
+{   
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "Tensor {{")?;
+        write!(f, "  data: ")?;
+        
+        let mut acc = String::new();
+        let mut multi_index = vec![0; D];
+        self.display_recursive(&mut acc, 0, &mut multi_index);
+        write!(f, "{}", acc)?;
+        writeln!(f, ",")?;
+        writeln!(f, "  shape:   {:?},", self.dims())?;
+        writeln!(f, "  device:  {:?},", self.device())?;
+        writeln!(f, "  backend: {:?},", B::name())?;
+        writeln!(f, "  dtype:   {:?},", "int")?; // this is probably cheating
```

Comment:
@antimora @nathanielsimard Ok I think I have something that works, but still only for Int tensors. Is there a good way to get the dtype out of the Backend struct that I'm missing?

Reply:
You have to specify the kind as a generic parameter to the tensor struct.

```rust
impl<B: Backend, const D: usize, K: BasicOps<B>> Tensor<B, D, K> {
   // K::Elem is an associative type, maybe we will it to implement core::fmt::Display
}
```
see https://github.com/burn-rs/burn/blob/main/burn-tensor/src/tensor/api/base.rs for more details.